apigateway/src/main/kotlin/ru/zveron/apigateway/registry/ProtoDefinitionRegistry.kt

```diff
@@ -0,0 +1,96 @@
+import kotlinx.coroutines.reactive.awaitSingle
+import mu.KLogging
+import org.springframework.cloud.netflix.eureka.reactive.EurekaReactiveDiscoveryClient
+import org.springframework.stereotype.Component
+import java.util.concurrent.ConcurrentHashMap
+
+@Component
+class ProtoDefinitionRegistry(
+    private val channelRegistry: GrpcChannelRegistry,
+    private val eurekaClient: EurekaReactiveDiscoveryClient,
+) {
+
+    companion object : KLogging()
+
+    private val reflectionMethodDescr = MethodDescriptor.newBuilder<ServerReflectionRequest, ServerReflectionResponse>()
+        .setFullMethodName(
+            MethodDescriptor.generateFullMethodName(
+                "grpc.reflection.v1alpha.ServerReflection",
+                "ServerReflectionInfo"
+            )
+        )
+        .setRequestMarshaller(ProtoLiteUtils.marshaller(ServerReflectionRequest.getDefaultInstance()))
+        .setResponseMarshaller(ProtoLiteUtils.marshaller(ServerReflectionResponse.getDefaultInstance()))
+        .setType(MethodDescriptor.MethodType.UNARY)
+        .build()
+
+    private val serviceToFileDescriptorMap = ConcurrentHashMap<String, FileDescriptor>()
+
+    private val serviceToProtoFile = ConcurrentHashMap<String, String?>()
+
+    suspend fun getProtoFileDescriptor(serviceName: String, grpcServiceName: String) =
+        serviceToFileDescriptorMap["$grpcServiceName-$serviceName"] ?: createProtoFileDescriptor(
+            grpcServiceName,
+            serviceName
+        )
+
+    private suspend fun createProtoFileDescriptor(
+        protoServiceName: String,
+        serviceName: String,
+    ): FileDescriptor {
+        val protoFileName = serviceToProtoFile["$protoServiceName-$serviceName"]
+            ?: eurekaClient.getInstances(serviceName)
+                .awaitSingle()?.metadata?.get(protoServiceName)
+                .also {
+                    serviceToProtoFile["$protoServiceName-$serviceName"] = it
+                }
+            ?: error("No service file provided in metadata")
+        logger.debug { "Proto file path $protoFileName" }
+
+        val serverReflectionRequest = ServerReflectionRequest.newBuilder()
+            .setFileByFilename(protoFileName)
+            .build()
+
+        val channel = channelRegistry.getChannel(serviceName)
+
+        val response = ClientCalls.unaryRpc(channel, reflectionMethodDescr, serverReflectionRequest)
+
+        if (response.hasErrorResponse()) {
+            logger.error { response.errorResponse }
+            throw RuntimeException(response.errorResponse.errorMessage)
+        }
+
+        val fileDescriptorProtos = response.fileDescriptorResponse.fileDescriptorProtoList.map {
+            DescriptorProtos.FileDescriptorProto.parseFrom(
+                it
+            )
+        }
+
+        val nameToProtoFile = fileDescriptorProtos.associateBy { it.name }
+
+        fileDescriptorProtos
+            .filter { it.name.equals(protoFileName, true) }
```

Comment:
я не понимаю что делает эта строчка, зачем нам сравнивать с названием прото файла.

Если мы ищем только одного конкретного контракта - то тут подойдет
```suggestion
            .find { it.name.equals(protoFileName, true) }
```

Reply:
чтобы получить именно его дескриптор, там помимо него еще другие могут лежать